### ISSUE TITLE
Supporting lossless handling of all raster pixel bit types

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2491]
tl;dr - This ticket is about enabling Mapnik's datasource plugins, image reader/writers, and raster-colorizer to better handle all raster pixel types. This is most relevant for allowing Mapnik to encode and decode more pixel types inside vector tiles.
### Background

Currently rendering raster data in Mapnik works like:
- Data is normally read with the [gdal.input plugin](https://github.com/mapnik/mapnik/blob/master/plugins/input/gdal/)
- Data is assumed to be RGBA by default and all bands are queried in order to assemble the assumed rgba data (https://github.com/mapnik/mapnik/blob/3cbde2dc547859c9cee3737a1820368530b047b5/plugins/input/gdal/gdal_featureset.cpp#L228-L438)
- Optionally a user can request to read a single band of data by passing `band=N` to the GDAL plugin which triggers reinterpreting the data as a 32 bit float: https://github.com/mapnik/mapnik/blob/3cbde2dc547859c9cee3737a1820368530b047b5/plugins/input/gdal/gdal_featureset.cpp#L210-L227
- Data can be styled on the fly using the `raster-colorizer` : https://github.com/mapnik/mapnik/blob/3cbde2dc547859c9cee3737a1820368530b047b5/include/mapnik/renderer_common/process_raster_symbolizer.hpp#L51-L53 but this only works right if the data represents a single band / float32.
### Limitations

This approach works but has a number of limitations:
- `reinterpret_cast` is bad practice / not type safe / hard to understand what is going on
- in the case of raw data being unsigned or integer, treating as floats is technically incorrect, potentially lossy, and may overflow in the case of 64 bit data.
- ideally the band could be requested via the styling language rather than being a datasource option

But, the real driving reasons being this ticket are:
- To provide a solid foundation for a [vector tile api](https://github.com/mapbox/mapnik-vector-tile) Mapnik needs to be able to losslessly encode single band data read originally from `gdal.input` inside a vector tile
- Then Mapnik needs be able to decode this single band data from within a vector tile so that:
  - the raster-colorizer can be used on that raw data to style it into rgab
  - The Mapnik api can be used to pass the data along to other vector tile clients that might wish to query or analyze the pixels

This is not yet possible because Mapnik image readers currently only know how to work with `rgba` for both [encoding](https://github.com/mapbox/mapnik-vector-tile/blob/5394eed6f3bcbbc16337414f08b054612b2c27c4/src/vector_tile_processor.hpp#L322) and [decoding](https://github.com/mapbox/mapnik-vector-tile/blob/5394eed6f3bcbbc16337414f08b054612b2c27c4/src/vector_tile_datasource.hpp#L149).
### Steps to implement

So, to get a pipeline built to support passing through, clipped but lossless pixel values I imagine this set of work:
- [x] 1) Store a `mapnik::util::variant<rgba,int16,int32,int64,float8,float32,...>` type on [mapnik::raster instead of the current hardcoded 32 bit rgba image](https://github.com/mapnik/mapnik/blob/3cbde2dc547859c9cee3737a1820368530b047b5/include/mapnik/raster.hpp#L39)
- [x] 2) Template `raster_colorizer::colorize` on pixel type: https://github.com/mapnik/mapnik/blob/4a2735fe91da0f762f2e8a96504c84014924561c/src/raster_colorizer.cpp#L125-L144. Perhaps rgba would be a no-op and only float/int types would attempt to be colorized?
- [x] 3) Template `scale_image_agg` to operate on either premultiplied rgba pixels (like it does now) or gray32 or gray64 pixels (refs #2585): https://github.com/mapnik/mapnik/blob/4a2735fe91da0f762f2e8a96504c84014924561c/src/image_scaling.cpp#L97-L111
- [x] 4) First phase: focus on tiff format only. So, when single band data is detected we would encode as tiff [here](https://github.com/mapbox/mapnik-vector-tile/blob/5394eed6f3bcbbc16337414f08b054612b2c27c4/src/vector_tile_processor.hpp#L322) instead of allowing webp/png/jpeg.
- [x] 5) To enable the above: enhance the [tiff reader](https://github.com/mapnik/mapnik/blob/33c1d21b7fe5f87d25bcc2651151e6b0391ae36b/src/tiff_reader.cpp) to detect and respect `TIFFTAG_BITSPERSAMPLE` and `TIFFTAG_PHOTOMETRIC` and the [writer](https://github.com/mapnik/mapnik/blob/33c1d21b7fe5f87d25bcc2651151e6b0391ae36b/include/mapnik/tiff_io.hpp) to be able to write them sensitive to the input image type.

After these changes are in place in mapnik core + mapnik-vector-tile then support would need to be added to node-mapnik as well in order to expose a javascript interface to single band rasters.

Longer term, once image readers support reading single band data, then we could unlock support for reading multiband data via the `raster.input` plugin.
